src/petals/client/routing/sequence_info.py

```diff
@@ -0,0 +1,104 @@
+
+use_hivemind_log_handler("in_root_logger")
+logger = get_logger(__file__)
+
+
+T = TypeVar("T")
+
+
+@dataclasses.dataclass
+class RemoteSequenceInfo:
+    """
+    A dataclass that stores general information about which servers hold any given layer;
+    - updated by RemoteSequenceManager in a background thread
+    - accessed by routing strategies in .on_update
+    :note: this class should *not* be modified by RoutingStrategy.on_update to avoid interference between strategies;
+     Any metadata specific to one routing strategy, it should be stored inside that strategy. Any information that
+     is used by most routing strategies should be moved from said strategies to this class.
+    """
+
+    block_uids: Tuple[ModuleUID, ...]
+    block_infos: Tuple[RemoteModuleInfo, ...]  # note: the contents of RemoteModuleInfo can and will be updated
+    spans_by_priority: List[RemoteSpanInfo]
+    spans_containing_block: Tuple[List[RemoteSpanInfo], ...]
+    last_updated_time: float
+
+    @classmethod
+    def make_empty(cls: Type[T], block_uids: Iterable[ModuleUID]) -> T:
+        block_uids = tuple(block_uids)
+        empty_block_infos = tuple(RemoteModuleInfo(uid) for uid in block_uids)
+        empty_spans = tuple([] for _ in range(len(block_uids)))
+        return cls(block_uids, empty_block_infos, [], empty_spans, last_updated_time=-float("inf"))
+
+    def __getitem__(self, ix: slice):
+        assert isinstance(ix, slice)
+        block_uids, block_infos = self.block_uids[ix], self.block_infos[ix]
+        spans_by_priority, spans_containing_block = self.compute_spans(block_infos)
+        return RemoteSequenceInfo(
+            block_uids, block_infos, spans_by_priority, spans_containing_block, self.last_updated_time
+        )
+
+    def __len__(self):
+        return len(self.block_uids)
+
+    def update_(self, new_block_infos: List[Optional[RemoteModuleInfo]]):
+        assert len(new_block_infos) == len(self.block_uids)
+        for block_index, (uid, info) in enumerate(zip(self.block_uids, new_block_infos)):
+            if info is None:
+                logger.debug(f"Found no block info for block {uid}")
+                continue
+            if not isinstance(info, RemoteModuleInfo):
+                logger.warning(f"Unexpected dht entry type for {uid}: {info}")
+                continue
+            if not info.servers:
+                logger.debug(f"Found no active peers for block {uid}")
+                continue
+            if info.uid != uid:
+                logger.warning(f"The DHT entry for {uid} actually points to {info.uid}")
+                continue
+            for server_id, (server_info, expiration_time) in info.servers.items():
+                self.block_infos[block_index].servers.store(server_id, server_info, expiration_time)
+
+        self.spans_by_priority, self.spans_containing_block = self.compute_spans(self.block_infos)
+        self.last_updated_time = time.perf_counter()
```

Comment:
```suggestion
        self.last_updated_time = get_dht_time()
```

`perf_counter()` has undefined offset.

Reply:
i respectfully disagree: this time is only used for computing differences, there is no offset involved